org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacProblemConverter.java

```diff
@@ -13,36 +13,94 @@
 
 package org.eclipse.jdt.internal.javac;
 
+import java.io.IOException;
 import java.util.Locale;
 
 import javax.tools.Diagnostic;
 import javax.tools.JavaFileObject;
 
+import org.eclipse.core.runtime.ILog;
 import org.eclipse.jdt.core.compiler.IProblem;
 import org.eclipse.jdt.internal.compiler.problem.ProblemSeverities;
 
 import com.sun.tools.javac.code.Kinds;
 import com.sun.tools.javac.code.Symbol;
+import com.sun.tools.javac.tree.JCTree.JCClassDecl;
+import com.sun.tools.javac.util.DiagnosticSource;
 import com.sun.tools.javac.util.JCDiagnostic;
+import com.sun.tools.javac.util.Position;
 
 public class JavacProblemConverter {
 	public static JavacProblem createJavacProblem(Diagnostic<? extends JavaFileObject> diagnostic) {
+		int problemId = toProblemId(diagnostic);
+		return switch (problemId) {
+			case IProblem.MissingSerialVersion -> missingSerialVersion(diagnostic, problemId);
+			default -> defaultProblem(diagnostic, problemId);
+		};
+	}
+
+	private static JavacProblem defaultProblem(Diagnostic<? extends JavaFileObject> diagnostic, int problemId) {
+		return new JavacProblem(
+				diagnostic.getSource().getName().toCharArray(),
+				diagnostic.getMessage(Locale.getDefault()),
+				diagnostic.getCode(),
+				problemId,
+				new String[0],
+				toSeverity(diagnostic),
+				(int) Math.min(diagnostic.getPosition(), diagnostic.getStartPosition()),
+				(int) (diagnostic.getEndPosition() - 1),
+				(int) diagnostic.getLineNumber(),
+				(int) diagnostic.getColumnNumber());
+	}
+
+	private static int toSeverity(Diagnostic<? extends JavaFileObject> diagnostic) {
+		return switch (diagnostic.getKind()) {
+			case ERROR -> ProblemSeverities.Error;
+			case WARNING, MANDATORY_WARNING -> ProblemSeverities.Warning;
+			case NOTE -> ProblemSeverities.Info;
+			default -> ProblemSeverities.Error;
+		};
+	}
+
+	private static JavacProblem missingSerialVersion(Diagnostic<? extends JavaFileObject> diagnostic, int problemId) {
```

Comment:
there's nothing specific to missing Serial Version UID in this method. It needs a more generic name (like classLevelProblem)

Reply:
Fixed.